rust/rope/src/find.rs

```diff
@@ -162,51 +176,116 @@ fn find_core(cursor: &mut Cursor<RopeInfo>, pat: &str,
-pub fn compare_cursor_str(cursor: &mut Cursor<RopeInfo>, mut pat: &str) -> bool {
+/// position on failure, but the end of the string on success. Returns the
+/// start position of the match.
+pub fn compare_cursor_str(cursor: &mut Cursor<RopeInfo>, mut pat: &str) -> Option<usize> {
+    let start_position = cursor.pos();
     if pat.is_empty() {
-        return true;
+        return Some(start_position);
     }
     let success_pos = cursor.pos() + pat.len();
     while let Some((leaf, pos_in_leaf)) = cursor.get_leaf() {
         let n = min(pat.len(), leaf.len() - pos_in_leaf);
         if leaf.as_bytes()[pos_in_leaf..pos_in_leaf + n] != pat.as_bytes()[..n] {
-            return false;
+            return None;
         }
         pat = &pat[n..];
         if pat.is_empty() {
             cursor.set(success_pos);
-            return true;
+            return Some(start_position);
         }
         let _ = cursor.next_leaf();
     }
-    false
+    None
 }
 
 /// Like `compare_cursor_str` but case invariant (using to_lowercase() to
-/// normalize both strings before comparison).
-fn compare_cursor_str_casei(cursor: &mut Cursor<RopeInfo>, pat: &str) -> bool {
+/// normalize both strings before comparison). Returns the start position
+/// of the match.
+fn compare_cursor_str_casei(cursor: &mut Cursor<RopeInfo>, pat: &str) -> Option<usize> {
+    let start_position = cursor.pos();
     let mut pat_iter = pat.chars();
     let mut c = pat_iter.next().unwrap();
     loop {
         if let Some(rope_c) = cursor.next_codepoint() {
             for lc_c in rope_c.to_lowercase() {
                 if c != lc_c {
-                    return false;
+                    return None;
                 }
                 if let Some(next_c) = pat_iter.next() {
                     c = next_c;
                 } else {
-                    return true;
+                    return Some(start_position);
                 }
             }
         } else {
             // end of string before pattern is complete
-            return false;
+            return None;
         }
     }
 }
 
+/// Compare whether the substring beginning at the cursor location matches
+/// the provided regular expression. The substring begins at the beginning
+/// of the leaf/start of the line.
+/// If the regular expression can match multiple lines then all leaves are
+/// consumed and matched against the regular expression. Otherwise only the
+/// current leaf is matched. Returns the start position of the match.
+fn compare_cursor_regex(cursor: &mut Cursor<RopeInfo>, pat: &str) -> Option<usize> {
+    let orig_position = cursor.pos();
+
+    if pat.is_empty() {
+        return Some(orig_position);
+    }
+
+    // create regex from untrusted input
+    match RegexBuilder::new(pat).size_limit(REGEX_SIZE_LIMIT).build() {
+        Ok(regex) => {
+            let mut text = String::new();
+
+            match cursor.get_leaf() {
```

Comment:
So this will work in the common case, but isn't fully correct - there's no guarantee that a line won't be fragmented between multiple leaves. It will definitely happen when lines are long (and if there's direct access to leaves, that case should probably be reflected in tests).

My sense of the best thing to do here is to use a lines iterator, which will abstract over the details of leaves, while being efficient and not allocating when line boundaries line up with leaf boundaries.

I think that means that you don't want to use the same Cursor<RopeInfo> for both string match and regex match; it's fine for string match because you can do a string match on a line fragment, so the efficiency gain in dealing with individual leaves is probably worth it.

Reply:
I was thinking we could use [`lines_raw`](https://docs.rs/xi-rope/0.2.0/src/xi_rope/rope.rs.html#318), which is probably going to be more efficient than a Cursor-based approach if it's strictly doing a forward scan. Is there some reason that won't work?

Reply:
Thinking about this, one difference between Cursor and LinesRaw is that in the latter case the client will have to keep track of the offset. But that can be done by adding up `line.len()` as we go along.

I can also see how to make a `Cursor` work, but it would at the very least require duplicating some of the logic we already have implemented in LinesRaw.

Reply:
I updated this pull request so that for regex search a line iterator is used. I am not sure if that is the best way to implement it.